File: app/api/routes/logs.py
from fastapi import APIRouter, Query
from sqlmodel import func, select

from app.api.deps import CurrentUser, SessionDep
from app.models import EventLog
from app.schemas import EventLogsPublic

router = APIRouter(tags=["logs"])


@router.get("/", response_model=EventLogsPublic)
def read_logs(
    session: SessionDep,
    current_user: CurrentUser,
    skip: int = Query(default=0, ge=0),
    limit: int = Query(default=100, ge=1, le=300),
    severity: str | None = Query(default=None),
    device_kind: str | None = Query(default=None),
    q: str | None = Query(default=None),
) -> EventLogsPublic:
    del current_user
    statement = select(EventLog)
    count_stmt = select(func.count()).select_from(EventLog)

    if severity:
        statement = statement.where(EventLog.severity == severity.lower())
        count_stmt = count_stmt.where(EventLog.severity == severity.lower())
    if device_kind:
        statement = statement.where(EventLog.device_kind == device_kind.lower())
        count_stmt = count_stmt.where(EventLog.device_kind == device_kind.lower())
    if q:
        pattern = f"%{q}%"
        statement = statement.where(
            (EventLog.message.ilike(pattern))
            | (EventLog.device_name.ilike(pattern))
            | (EventLog.ip_address.ilike(pattern))
            | (EventLog.event_type.ilike(pattern))
        )
        count_stmt = count_stmt.where(
            (EventLog.message.ilike(pattern))
            | (EventLog.device_name.ilike(pattern))
            | (EventLog.ip_address.ilike(pattern))
            | (EventLog.event_type.ilike(pattern))
        )

    count = session.exec(count_stmt).one()
    logs = session.exec(statement.order_by(EventLog.created_at.desc()).offset(skip).limit(limit)).all()
    return EventLogsPublic(data=logs, count=count)

File: frontend/src/api/logs.ts
import api from "./http";

export type EventSeverity = "info" | "warning" | "error" | "critical";
export type EventDeviceKind = "printer" | "media_player" | "switch" | "cash_register" | null;

export interface EventLog {
  id: string;
  severity: EventSeverity;
  category: string;
  event_type: string;
  message: string;
  device_kind: EventDeviceKind;
  device_name: string | null;
  ip_address: string | null;
  created_at: string;
}

export interface EventLogsResponse {
  data: EventLog[];
  count: number;
}

export async function getEventLogs(params?: {
  skip?: number;
  limit?: number;
  severity?: EventSeverity | "all";
  device_kind?: "printer" | "media_player" | "switch" | "cash_register" | "all";
  q?: string;
}) {
  const query: Record<string, string | number> = {
    skip: params?.skip ?? 0,
    limit: params?.limit ?? 100,
  };
  if (params?.severity && params.severity !== "all") query.severity = params.severity;
  if (params?.device_kind && params.device_kind !== "all") query.device_kind = params.device_kind;
  if (params?.q) query.q = params.q;
  const { data } = await api.get<EventLogsResponse>("/logs/", { params: query });
  return data;
}
